valuable-serde/src/lib.rs

```diff
@@ -0,0 +1,485 @@
+        serializer: S,
+    },
+    End(Result<S::Ok, S::Error>),
+    Tmp,
+}
+
+impl<S: Serializer> Visit for VisitStaticStruct<S> {
+    fn visit_named_fields(&mut self, named_values: &NamedValues<'_>) {
+        let (name, fields, serializer) = match mem::replace(self, Self::Tmp) {
+            Self::Start {
+                name,
+                fields: Fields::Named(fields),
+                serializer,
+            } => (name, fields, serializer),
+            _ => unreachable!(),
+        };
+        let mut ser = match serializer.serialize_struct(name, named_values.len()) {
+            Ok(ser) => ser,
+            Err(e) => {
+                *self = Self::End(Err(e));
+                return;
+            }
+        };
+        for (i, (_, v)) in named_values.entries().enumerate() {
+            if let Err(e) = ser.serialize_field(fields[i].name(), &SerdeValue(v.as_value())) {
+                *self = Self::End(Err(e));
+                return;
+            }
+        }
+        *self = Self::End(ser.end());
+    }
+
+    fn visit_unnamed_fields(&mut self, values: &[Value<'_>]) {
+        let (name, serializer) = match mem::replace(self, Self::Tmp) {
+            Self::Start {
+                name,
+                fields: Fields::Unnamed,
+                serializer,
+            } => (name, serializer),
+            _ => unreachable!(),
+        };
+        if values.len() == 1 {
+            *self = Self::End(serializer.serialize_newtype_struct(name, &SerdeValue(values[0])));
+            return;
+        }
+        let mut ser = match serializer.serialize_tuple_struct(name, values.len()) {
+            Ok(ser) => ser,
+            Err(e) => {
+                *self = Self::End(Err(e));
+                return;
+            }
+        };
+        for v in values {
+            if let Err(e) = ser.serialize_field(&SerdeValue(v.as_value())) {
+                *self = Self::End(Err(e));
+                return;
+            }
+        }
+        *self = Self::End(ser.end());
+    }
+}
+
+// Dynamic struct will be serialized as map or seq.
+enum VisitDynamicStruct<S: serde::Serializer> {
+    Start(S),
+    End(Result<S::Ok, S::Error>),
+    Tmp,
+}
+
+impl<S: serde::Serializer> Visit for VisitDynamicStruct<S> {
+    fn visit_named_fields(&mut self, named_values: &NamedValues<'_>) {
+        let serializer = match mem::replace(self, Self::Tmp) {
+            Self::Start(serializer) => serializer,
+            _ => unreachable!(),
+        };
+        let mut ser = match serializer.serialize_map(Some(named_values.len())) {
+            Ok(ser) => ser,
+            Err(e) => {
+                *self = Self::End(Err(e));
+                return;
+            }
+        };
+        for (f, v) in named_values.entries() {
+            if let Err(e) = ser.serialize_entry(f.name(), &SerdeValue(v.as_value())) {
+                *self = Self::End(Err(e));
+                return;
+            }
+        }
+        *self = Self::End(ser.end());
```

Comment:
We need to consider the possibility that `visit_named_fields`/`visit_unnamed_fields` will be called multiple times? If so, I will need to fix this implementation.
```rust
impl Valuable for S {
    fn as_value(&self) -> Value<'_> {
        Value::Structable(self)
    }
    fn visit(&self, v: &mut dyn Visit) {
        v.visit_named_fields(&NamedValues::new(
            WORLD_FIELDS,
            &[...],
        ));
        v.visit_named_fields(&NamedValues::new(
            WORLD_FIELDS2,
            &[...],
        ));
    }
}
```

Reply:
Yes, it can be called multiple times when the struct/enum is *dynamic*.

Reply:
Fixed (https://github.com/tokio-rs/valuable/pull/23/commits/1421d89f2c697fc24c3db09ea337b50a4ae1b50a).